utils/retrieve-state/src/lib.rs

```diff
@@ -714,3 +721,227 @@ fn spawn_peer(
+
+    // Channel for a worker thread to send an error.
+    let (err_tx, mut err_rx) = mpsc::channel(1);
+
+    let queue = Arc::new(WorkQueue::default());
+    queue.push_job(state_root_hash);
+    let max_parallel_trie_fetches = max_parallel_trie_fetches.min(peers.len());
+    info!(
+        "using {}/{} valid peers",
+        max_parallel_trie_fetches,
+        peers.len()
+    );
+    let workers: FuturesUnordered<_> = peers[0..max_parallel_trie_fetches]
+        .iter()
+        .enumerate()
+        .map(|(worker_id, address)| {
+            info!("spawning worker {} for {:?}", worker_id, address);
+            tokio::spawn(sync_trie_store_worker(
+                worker_id,
+                err_tx.clone(),
+                queue.clone(),
+                engine_state.clone(),
+                client.clone(),
+                *address,
+                Arc::clone(&bps_counter),
+            ))
+        })
+        .collect();
+    drop(err_tx);
+    workers.for_each(|_| async move {}).await;
+    if let Some(err) = err_rx.recv().await {
+        return Err(err); // At least one download failed: return the error.
+    }
+
+    Ok(())
+}
+
+/// A worker task that takes trie keys from a queue and downloads the trie.
+async fn sync_trie_store_worker(
+    worker_id: usize,
+    err_tx: mpsc::Sender<anyhow::Error>,
+    queue: Arc<WorkQueue<Digest>>,
+    engine_state: Arc<EngineState<LmdbGlobalState>>,
+    client: Client,
+    address: SocketAddr,
+    bps_counter: Arc<Mutex<(Instant, usize)>>,
+) {
+    while let Some(job) = queue.next_job().await {
+        if err_tx.capacity() < 1 {
+            return; // Another task failed and sent an error.
+        }
+        // Calculate global (all workers) total of trie bytes decoded per second.
+        {
+            let (last_update, total_bytes_since_update) = &mut *bps_counter.lock().await;
+            let elapsed = last_update.elapsed().as_millis();
+            if elapsed >= 1000 {
+                let scaled = *total_bytes_since_update as f32 * (elapsed as f32 / 1000f32);
+                info!("throughput rate: {} kbps", scaled / 1024f32);
+                *total_bytes_since_update = 0;
+                *last_update = Instant::now();
+            }
+        }
+        trace!(worker_id, trie_key = %job.inner(), "worker downloading trie");
+        match fetch_and_store_trie(
+            engine_state.clone(),
+            &client,
+            address,
+            *job.inner(),
+            bps_counter.clone(),
+        )
+        .await
+        {
+            Ok(Some(child_jobs)) => {
+                for child_job in child_jobs {
+                    queue.push_job(child_job);
+                }
+            }
+            Ok(None) => {
+                warn!(
+                    "got no bytes back for trie, requeuing job and killing worker {}",
+                    worker_id
+                );
+                queue.push_job(*job.inner());
+                return;
```

Comment:
Why kill the worker in that case?

Reply:
In my tests, keeping the worker around led to a lot more attempts to request other tries from the same worker that failed with zero bytes.

Reply:
Ah, I just realized that each worker corresponds to one peer.
Isn't that inefficient? We should easily be able to make multiple parallel requests to the same peer.

Reply:
Have you tried running the tool?